go/adbc/driver/databricks/cmd_record_reader.go

```diff
@@ -0,0 +1,198 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package databricks
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	"github.com/apache/arrow-adbc/go/adbc"
+	"github.com/apache/arrow-go/v18/arrow"
+	"github.com/databricks/databricks-sdk-go/service/compute"
+)
+
+type commandReader struct {
+	refCount int64
+
+	cmdExecution compute.CommandExecutionInterface
+
+	// Command Execution that this reader is associated with.
+	CommandId string
+
+	Results *compute.Results
+
+	rec      arrow.Record
+	err      error
+	schema   *arrow.Schema
+
+	cancelFn context.CancelFunc
+}
+
+func DeriveSchema(dbx_schema []map[string]interface{}) (*arrow.Schema, error) {
+	fields := make([]arrow.Field, len(dbx_schema))
+	for i, col := range dbx_schema {
+		var arrowType arrow.DataType
+		colType := strings.Trim(col["type"].(string), "\"")
+		switch {
+			case colType == "int":
+				arrowType = arrow.PrimitiveTypes.Int32
+			case colType == "integer":
+				arrowType = arrow.PrimitiveTypes.Int32
+			case colType == "string":
+				arrowType = arrow.BinaryTypes.String
+			case colType == "short":
+				arrowType = arrow.PrimitiveTypes.Int16
+			case colType == "long":
+				arrowType = arrow.PrimitiveTypes.Int64
+			case colType == "float":
+				arrowType = arrow.PrimitiveTypes.Float32
+			case colType == "double":
+				arrowType = arrow.PrimitiveTypes.Float64
+			case colType == "boolean":
+				arrowType = arrow.FixedWidthTypes.Boolean
+			case colType == "timestamp":
+				arrowType = &arrow.TimestampType{Unit: arrow.Second}
+			case colType == "date":
+				arrowType = arrow.FixedWidthTypes.Date32
+			case strings.HasPrefix(colType, "decimal"):
+				// Parse decimal precision and scale from format "decimal(precision,scale)"
+				if matches := regexp.MustCompile(`decimal\((\d+),(\d+)\)`).FindStringSubmatch(colType); matches != nil {
+					precision, _ := strconv.ParseInt(matches[1], 10, 32)
+					scale, _ := strconv.ParseInt(matches[2], 10, 32)
+					arrowType = &arrow.Decimal256Type{Precision: int32(precision), Scale: int32(scale)}
```

Comment:
Let default to `arrow.Decimal128Type` and use `Decimal256Type` only if precision is greater than 38.

This is because `2^128 >= 38^10`. Proof:

```
log10(2^128) >= log10(38^10)
log10(2^128) >= 38
38.531839445 >= 38
```

128 bits can store any number of 38 decimal digits. Scale controls where the `.` goes so it doesn't have to be considered in this calculation.